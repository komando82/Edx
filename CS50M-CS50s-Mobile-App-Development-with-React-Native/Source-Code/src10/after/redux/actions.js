import {login} from '../api'

// action types
export const UPDATE_USER = 'UPDATE_USER'
export const UPDATE_CONTACT = 'UPDATE_CONTACT'
export const LOG_IN_SENT = 'LOG_IN_SENT'
export const LOG_IN_FULFILLED = 'LOG_IN_FULFILLED'
export const LOG_IN_REJECTED = 'LOG_IN_REJECTED'

// action creators
export const updateUser = update => ({
  type: UPDATE_USER,
  payload: update,
})

export const addContact = newContact => ({
  type: UPDATE_CONTACT,
  payload: newContact,
})

// async action creator
export const logInUser = (username, password) => async dispatch => {
  dispatch({type: LOG_IN_SENT})
  try {
    const token = await login(username, password)
    dispatch({type: LOG_IN_FULFILLED, payload: token})
  } catch (err) {
    dispatch({type: LOG_IN_REJECTED, payload: err.message})
  }
}
