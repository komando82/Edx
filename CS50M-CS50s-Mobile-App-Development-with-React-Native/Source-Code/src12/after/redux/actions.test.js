/* global jest, expect, describe, it */
/* eslint-disable */

import * as actions from './actions'

describe('updateUser returns actions', () => {
  it('returns an action', () => {
    expect(actions.updateUser({name: 'test name'})).toMatchSnapshot()
  })

  it('handles empty object', () => {
    expect(actions.updateUser({})).toMatchSnapshot()
  })

  it('handles empty name', () => {
    expect(actions.updateUser({name: ''})).toMatchSnapshot()
  })
})

describe('logInUser returns actions', () => {
  const errMessage = 'incorrect creds'
  const fakeToken = 'thisIsATestToken'
  const mockLogin = (username, password) => {
    if (username === 'u' && password === 'p') {
      return fakeToken
    }
    throw new Error(errMessage)
  }

  it('dispatches LOG_IN_SENT', async () => {
    const mockDispatch = jest.fn()
    await actions.logInUser('', '')(mockDispatch)
    // mockDispatch.mock.calls all the args that the mock fn was invoked on
    expect(mockDispatch.mock.calls[0][0]).toEqual({type: actions.LOG_IN_SENT})
  })

  it('dispatches LOG_IN_FULFILLED with correct creds', async () => {
    const mockDispatch = jest.fn()
    await actions.logInUser('u', 'p', mockLogin)(mockDispatch)

    expect(mockDispatch.mock.calls[1][0]).toEqual({type: actions.LOG_IN_FULFILLED, payload: fakeToken})
    expect(mockDispatch.mock.calls[1]).toMatchSnapshot()
  })

  it('dispatches LOG_IN_REJECTED with incorrect creds', async () => {
    const mockDispatch = jest.fn()
    await actions.logInUser('', '', mockLogin)(mockDispatch)

    expect(mockDispatch.mock.calls[1][0]).toEqual({type: actions.LOG_IN_REJECTED, payload: errMessage})
    expect(mockDispatch.mock.calls[1]).toMatchSnapshot()
  })
})
