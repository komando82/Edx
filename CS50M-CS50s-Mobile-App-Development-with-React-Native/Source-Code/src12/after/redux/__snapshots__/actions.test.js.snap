// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logInUser returns actions dispatches LOG_IN_FULFILLED with correct creds 1`] = `
Array [
  Object {
    "payload": "thisIsATestToken",
    "type": "LOG_IN_FULFILLED",
  },
]
`;

exports[`logInUser returns actions dispatches LOG_IN_REJECTED with incorrect creds 1`] = `
Array [
  Object {
    "payload": "incorrect creds",
    "type": "LOG_IN_REJECTED",
  },
]
`;

exports[`updateUser returns actions handles empty name 1`] = `
Object {
  "payload": Object {
    "name": "",
  },
  "type": "UPDATE_USER",
}
`;

exports[`updateUser returns actions handles empty object 1`] = `
Object {
  "payload": Object {},
  "type": "UPDATE_USER",
}
`;

exports[`updateUser returns actions returns an action 1`] = `
Object {
  "payload": Object {
    "name": "test name",
  },
  "type": "UPDATE_USER",
}
`;
